### DISCUSSION TITLE
[week20 優化網站] webpack bundle 問題

### DISCUSSION THREAD

**Discussion by EshauHuang** [Lidemy/forum#207] [前端相關問題]
在此先感謝各位解答！

目前因為 final project 需要用到大量的圖片，因此想先學習如何優化有大量圖片的網站，所以正照著 [yakim-shu大大的網站優化成果](https://github.com/Lidemy/lazy-hackathon/issues/7) 優化範例網站。但目前遇到兩個問題：
* 無法壓縮 css(scss) 檔案
* 使用 webpack 匯入 jquery 某些功能沒辦法使用

### 無法壓縮 css(scss) 檔案
這是一開始的設定檔，為了讓 css 內的 url 能夠讀到圖片，因此使用 `url-loader`
```js
// webpack.config.js
module.exports = {
  entry: "./src/index.js",
  output: {
    filename: "bundle.js",
    path: path.resolve(__dirname, "dist"),
  },
  module: {
    rules: [
      {
        test: /\.m?js$/,
        exclude: /node_modules/,
        use: [
          {
            loader: "babel-loader",
            options: {
              presets: ["@babel/preset-env"],
            },
          },
        ],
      },
      {
        test: "/.(png|jpg)$/",
        loader: "url-loader",
      },
      {
        test: /\.css$/,
        use: ["style-loader", "css-loader"],
      },
    ],
  },
};

```
但執行 build 後會出現錯誤訊息 `Module not found: Error: Can't resolve './../image/'`，所以猜測應該就是文章提到的「要記得加上 SCSS 前綴，不然 CSS 的 mask 語法會出錯」，而當我把網址是 `./../image/` 給註解掉，就能成功壓縮。

嘗試使用 yakim-shu 大大提到的 `gulp-postcss` 與 `autoprefixer`，不過因為是使用 webpack 所以找了些文章參考。

參考：**[webpack文件-postcss-loader](https://webpack.js.org/loaders/postcss-loader/)**
```js
// webpack.config.js
rules: [
      {
        test: /\.css$/i,
        use: [
          "style-loader",
          "css-loader",
          {
            loader: "postcss-loader",
            options: {
              postcssOptions: {
                plugins: [
                  [
                    "postcss-preset-env",
                    {
                      // Options
                    },
                  ],
                ],
              },
            },
          },
        ],
      },
    ],
  }

```
參考：**[在webpack中使用postcss 和 autoprefixer幫css自動增加前綴詞](https://medium.com/%E5%B0%8F%E5%BD%A5%E5%BD%A5%E7%9A%84%E5%89%8D%E7%AB%AF%E4%BA%94%E5%9B%9B%E4%B8%89/%E5%9C%A8webpack%E4%B8%AD%E4%BD%BF%E7%94%A8postcss-%E5%92%8C-autoprefixer%E5%B9%ABcss%E8%87%AA%E5%8B%95%E5%A2%9E%E5%8A%A0%E5%89%8D%E7%B6%B4%E8%A9%9E-c06c093dc60e)**
```js
// 新增 postcss.config.js
module.exports = {
  plugins: [
    require("autoprefixer")({
      grid: true,
      overrideBrowserslist: [
        "> 1%",
        "last 5 versions",
        "Firefox >= 45",
        "ios >= 8",
        "ie >= 10",
      ],
    }),
  ],
};

// webpack.config.js
rules: [
      {
        test: /\.(sass|scss)$/,
        use: ['css-loader','postcss-loader','sass-loader']
      },
    ]
```
兩者皆會出現錯誤訊息 `Module parse failed: Unexpected character '@'`，雖然有做後續的爬文動作，但還是沒有找到有效的解決方式。

### 使用 webpack 匯入 jquery 某些功能沒辦法使用
* 直接在 index.js 匯入 jquery(使用內建的 jquery-3.4.1.js)
```js
// index.js
import * as $ from "./jquery-3.4.1.js";
import $ from "./jquery-3.4.1.js";
```
* 參考 [Managing jQuery plugin dependency in webpack](https://stackoverflow.com/questions/28969861/managing-jquery-plugin-dependency-in-webpack)，用 npm 安裝了 jquery，並使用了 ProvidePlugin 與 plugin 加入 jquery

```js
// index.js
import $ from "jquery";

// webpack.config.js
plugins: [
  new webpack.ProvidePlugin({
    $: "jquery",
    jQuery: "jquery",
  }),
],

```
兩者在點擊上方的「比賽簡介、參賽隊伍....」會沒有反應，且當我再次在 html 把 jquery 匯入還是會沒有反應。

再次感謝各位的解答，謝謝！

**Comment by aszx87410** [comment]
jQuery 那個，你用 npm 安裝之後應該就不用另外做設定了，你另外做設定應該反而是要引用外部的 jQuery 的意思

**Reply by EshauHuang** [reply]
感謝老師的糾正，誤會了 ProvidePlugin 的用法，後來再看了一次作者的文章才發現有提到「打包成 1 隻 js 會爆掉，我猜是各自有相依關係，但打包順序無法控制」，所以還是決定使用 <script> 引入 jquery，感謝老師。
